// @unocss-applet/preset-applet@0.2.18/dist/index.cjs
'use strict';

const presetWind = require('@unocss/preset-wind');
const presetMini = require('@unocss/preset-mini');
const core = require('@unocss/core');
const utils = require('@unocss/preset-mini/utils');

const preflights = [
  {
    layer: "preflights",
    getCSS(ctx) {
      if (ctx.theme.preflightBase)
        return `page,::before,::after{${core.entriesToCss(Object.entries(ctx.theme.preflightBase))}}`;
    }
  }
];

function unoCSSToAppletProcess(str) {
  if (str.includes("."))
    str = str.replace(/\\\./g, "-point-");
  if (str.includes("/"))
    str = str.replace(/\\\//g, "-div-");
  if (str.includes(":"))
    str = str.replace(/\\\:/g, "-c-");
  if (str.includes("%"))
    str = str.replace(/\\\%/g, "-pct");
  if (str.includes("!"))
    str = str.replace(/\\\!/g, "i-");
  if (str.includes("#"))
    str = str.replace(/\\\#/g, "-h-");
  if (str.includes("("))
    str = str.replace(/\\\(/g, "p-");
  if (str.includes(")"))
    str = str.replace(/\\\)/g, "-q");
  if (str.includes("["))
    str = str.replace(/\\\[/g, "l-");
  if (str.includes("]"))
    str = str.replace(/\\\]/g, "-r");
  if (str.includes("{"))
    str = str.replace(/\\\{/g, "c-");
  if (str.includes("}"))
    str = str.replace(/\\\}/g, "-c");
  if (str.includes(","))
    str = str.replace(/\\\,/g, "-comma-");
  if (str.includes("$"))
    str = str.replace(/\\\$/g, "-d-");
  return str;
}

const mixComponent = (v1, v2, w) => `calc(${v2} + (${v1} - ${v2}) * ${w} / 100)`;
const mixColor = (color1, color2, weight) => {
  const colors = [color1, color2];
  const cssColors = [];
  for (let c = 0; c < 2; ++c) {
    const color = typeof colors[c] === "string" ? utils.parseCssColor(colors[c]) : colors[c];
    if (!color || !["rgb", "rgba"].includes(color.type))
      return;
    cssColors.push(color);
  }
  const newComponents = [];
  for (let x = 0; x < 3; ++x)
    newComponents.push(mixComponent(cssColors[0].components[x], cssColors[1].components[x], weight));
  return {
    type: "rgb",
    components: newComponents,
    alpha: mixComponent(cssColors[0].alpha ?? 1, cssColors[1].alpha ?? 1, weight)
  };
};
const tint = (color, weight) => mixColor("#fff", color, weight);
const shade = (color, weight) => mixColor("#000", color, weight);
const shift = (color, weight) => {
  const num = parseFloat(`${weight}`);
  if (!Number.isNaN(num))
    return num > 0 ? shade(color, weight) : tint(color, -num);
};
const fns = { tint, shade, shift };
const variantColorMix = (matcher) => {
  const m = matcher.match(/^mix-(tint|shade|shift)-(-?\d{1,3})[-:]/);
  if (m) {
    return {
      matcher: matcher.slice(m[0].length),
      body: (body) => {
        body.forEach((v) => {
          if (v[1]) {
            const color = utils.parseCssColor(`${v[1]}`);
            if (color) {
              const mixed = fns[m[1]](color, m[2]);
              if (mixed)
                v[1] = utils.colorToString(mixed);
            }
          }
        });
        return body;
      }
    };
  }
};

function presetApplet(options = {}) {
  options.dark = options.dark ?? "class";
  options.attributifyPseudo = options.attributifyPseudo ?? false;
  options.preflight = options.preflight ?? true;
  const enable = options.enable ?? true;
  return {
    name: "unocss-preset-applet",
    theme: presetWind.theme,
    rules: presetWind.rules,
    shortcuts: presetWind.shortcuts,
    variants: [
      ...presetWind.variants(options),
      variantColorMix
    ],
    options,
    preflights: options.preflight ? enable ? preflights : presetMini.preflights : [],
    postprocess: [
      (util) => {
        enable && (util.selector = unoCSSToAppletProcess(util.selector));
        return util;
      }
    ],
    prefix: options.prefix
  };
}

module.exports = presetApplet;
